pallets/connectors-gateway/src/lib.rs

```diff
@@ -0,0 +1,273 @@
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub (super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::origin]
+	pub type Origin = GatewayOrigin;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The origin type.
+		type RuntimeOrigin: Into<Result<GatewayOrigin, <Self as frame_system::Config>::RuntimeOrigin>>
+			+ From<GatewayOrigin>;
+
+		/// The event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// The LocalOrigin ensures that some calls can only be performed from a
+		/// local context i.e. a different pallet.
+		type LocalOrigin: EnsureOrigin<
+			<Self as frame_system::Config>::RuntimeOrigin,
+			Success = DomainAddress,
+		>;
+
+		/// The AdminOrigin ensures that some calls can only be performed by
+		/// admins.
+		type AdminOrigin: EnsureOrigin<<Self as frame_system::Config>::RuntimeOrigin>;
+
+		/// The incoming and outgoing message type.
+		///
+		/// NOTE - this `Codec` trait is the Centrifuge trait for connectors
+		/// messages.
+		type Message: Codec;
+
+		/// The message router type that is stored for each domain.
+		type Router: DomainRouter<Sender = Self::AccountId, Message = Self::Message>
+			+ Clone
+			+ Debug
+			+ MaxEncodedLen
+			+ TypeInfo
+			+ FullCodec
+			+ EncodeLike
+			+ PartialEq;
+
+		/// The type that processes incoming messages.
+		type InboundQueue: InboundQueue<Sender = DomainAddress, Message = Self::Message>;
+
+		type WeightInfo: WeightInfo;
+
+		/// Maximum size of an incoming message.
+		#[pallet::constant]
+		type MaxIncomingMessageSize: Get<u32>;
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub (super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// The router for a given domain was set.
+		DomainRouterSet { domain: Domain, router: T::Router },
+
+		/// A connector was added to a domain.
+		ConnectorAdded { connector: DomainAddress },
+
+		/// A connector was removed from a domain.
+		ConnectorRemoved { connector: DomainAddress },
+	}
+
+	/// Storage for domain routers.
+	///
+	/// This can only be set by an admin.
+	#[pallet::storage]
+	pub(crate) type DomainRouters<T: Config> = StorageMap<_, Blake2_128Concat, Domain, T::Router>;
+
+	/// Storage that contains a limited number of whitelisted connectors for a
+	/// particular domain.
+	///
+	/// This can only be modified by an admin.
+	#[pallet::storage]
+	pub(crate) type ConnectorsAllowlist<T: Config> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		Domain,
+		Blake2_128Concat,
+		DomainAddress,
+		(),
+		ValueQuery,
+	>;
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Router initialization failed.
+		RouterInitFailed,
+
+		/// The origin of the message to be processed is invalid.
+		InvalidMessageOrigin,
+
+		/// The domain is not supported.
+		DomainNotSupported,
+
+		/// Message decoding error.
+		MessageDecode,
```

Comment:
Nit: Can we make this error more explicit like `InvalidIncomingMessage` or at least `MessageDecodingFailed`?

Reply:
Will do, thank you!